source-mysql-batch/main.go

```diff
@@ -0,0 +1,194 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+
+	schemagen "github.com/estuary/connectors/go/schema-gen"
+	boilerplate "github.com/estuary/connectors/source-boilerplate"
+	pf "github.com/estuary/flow/go/protocols/flow"
+	log "github.com/sirupsen/logrus"
+
+	_ "github.com/go-mysql-org/go-mysql/driver"
+)
+
+// Config tells the connector how to connect to and interact with the source database.
+type Config struct {
+	Address  string         `json:"address" jsonschema:"title=Server Address,description=The host or host:port at which the database can be reached." jsonschema_extras:"order=0"`
+	User     string         `json:"user" jsonschema:"default=flow_capture,description=The database user to authenticate as." jsonschema_extras:"order=1"`
+	Password string         `json:"password" jsonschema:"description=Password for the specified database user." jsonschema_extras:"secret=true,order=2"`
+	Advanced advancedConfig `json:"advanced,omitempty" jsonschema:"title=Advanced Options,description=Options for advanced users. You should not typically need to modify these." jsonschema_extra:"advanced=true"`
+	// TODO(wgd): Add network tunnel support
+}
+
+type advancedConfig struct {
+	DBName string `json:"dbname,omitempty" jsonschema:"title=Database Name,default=mysql,description=The name of database to connect to. In general this shouldn't matter. The connector can discover and capture from all databases it's authorized to access."`
+}
+
+// Validate checks that the configuration possesses all required properties.
+func (c *Config) Validate() error {
+	var requiredProperties = [][]string{
+		{"address", c.Address},
+		{"user", c.User},
+		{"password", c.Password},
+	}
+	for _, req := range requiredProperties {
+		if req[1] == "" {
+			return fmt.Errorf("missing '%s'", req[0])
+		}
+	}
+	return nil
+}
+
+// SetDefaults fills in the default values for unset optional parameters.
+func (c *Config) SetDefaults() {
+	// The address config property should accept a host or host:port
+	// value, and if the port is unspecified it should be the MySQL
+	// default 3306.
+	if !strings.Contains(c.Address, ":") {
+		c.Address += ":3306"
+	}
+}
+
+// ToDSN converts the Config to a DSN string.
+func (c *Config) ToDSN() string {
+	// go-mysql dsn format: "user:password@addr?dbname"
+	// For simplicity we use the URL struct and strip the leading 'mysql://'
+	// Ironically, the client library does the exact inverse to parse it.
+	var address = c.Address
+	var uri = url.URL{
+		Scheme: "mysql",
+		User:   url.UserPassword(c.User, c.Password),
```

Comment:
I haven't really dug into the larger question of which MySQL driver is better for our purposes here (the `go-mysql-org` one is what we use for the CDC connector because it has replication support, and I used it here because it's the one I've worked with before). But I happen to know that this client library uses `url.Parse` to decode the DSN back into host/user/password so _hopefully_ that round-trips correctly.